### DISCUSSION TITLE
Separate types for TypeProvider for validation and serialization.

### DISCUSSION THREAD

**Discussion by Bram-dc** [fastify/fastify#5311] [Ideas]
I am trying to improve the zod type provider to also work with .transform()

This is my current simplified code:
```
export interface ZodTypeProvider extends FastifyTypeProvider {
    output: this['input'] extends z.ZodTypeAny ? z.output<this['input']> : never
}

export const validatorCompiler: FastifySchemaCompiler<z.ZodTypeAny> = ({ schema }) => data => {
    const result = schema.safeParse(data)
    if (result.success)
        return { value: result.data }

    return { error: result.error }
}

export const serializerCompiler: FastifySerializerCompiler<z.ZodTypeAny> = ({ schema }) => data => {
    const result = schema.safeParse(data)
    if (result.success)
        return JSON.stringify(result.data)

    throw new Error(`Failed to create view: ${result.error.message}`)
}
```

I would like to pass the output from a route (the data that gets serialized) to be typed against the input of a zod schema. The types from querystring, params and body not however since the types are the output of the zod schema.

Zod has an z.input<T> and z.output<T> methods. But currently in fastify we can only create types for both at once.

**Would there be a way to create types separatly?**

Like for example:
```
export interface ZodValidatorTypeProvider extends FastifyValidatorTypeProvider {
    output: this['input'] extends z.ZodTypeAny ? z.output<this['input']> : never
}

export interface ZodSerializerTypeProvider extends FastifySerializerTypeProvider {
    output: this['input'] extends z.ZodTypeAny ? z.input<this['input']> : never
}
```

**Comment by Bram-dc** [comment]
This is the current TypeProvider template:
```
export interface FastifyTypeProvider {
  readonly input: unknown,
  readonly output: unknown,
}
```

Why is the schema even called "input"? Would it not make more sense to call it "schema"?
Then you could have seperate types like this:
```
export interface FastifyTypeProvider {
  readonly schema: unknown,
  readonly input: unknown,
  readonly output: unknown,
}
```

Or even more configurable:
```
export interface FastifyValidatorTypeProvider {
  readonly schema: unknown,
  readonly output: unknown,
}

export interface FastifySerializerTypeProvider {
  readonly schema: unknown,
  readonly input: unknown,
}
```

**Comment by mcollina** [comment]
I don't understand why you would add a validation against the _output_ of a validation. Your user would need to specify the input, not the output, at the route level.

Overall I'm not opposed to make things more flexible, but it would need not to be breaking.

**Reply by Bram-dc** [reply]
It is only to correctly type the return value of a route:

Here I get an error:
![image](https://github.com/fastify/fastify/assets/13683918/e0c2017f-07cf-4a58-aebc-78ac72ec0e74)

Here I do not:
![image](https://github.com/fastify/fastify/assets/13683918/679ecc95-a9f8-4b51-98a6-ce7891efb423)

But both ways will have the same output, it is just the type definition that is incorrect.

Because zod using the "coerce" property will convert a number to string anyways during the serialization.
I would like to correctly type this.

But not for the validation of params/querystring/body since in the handler we get the output of the validation.

**Reply by Bram-dc** [reply]
The value returned by the return keyword is the "input" of the serialization, after that it serializes that value using the "schema"

**Reply by mcollina** [reply]
I would review such a PR, if you want to contribute it.

**Comment by Bram-dc** [comment]
#5315
I think someone should review the naming of "input" and "output", but this PR should do the job.

**Reply by Bram-dc** [reply]
This is how a new TypeProvider, in this case for Zod, would look like. Without any breaking changes to the old ones.

```
interface ZodValidatorTypeProvider extends FastifyTypeProvider {
    output: this['input'] extends z.ZodTypeAny ? z.output<this['input']> : never
}

interface ZodSerializerTypeProvider extends FastifyTypeProvider {
    output: this['input'] extends z.ZodTypeAny ? z.input<this['input']> : never
}

export interface ZodTypeProvider extends FastifySeparatedTypeProvider {
    validator: ZodValidatorTypeProvider,
    serializer: ZodSerializerTypeProvider,
}
```

**Comment by Bram-dc** [comment]
Solution 1:
```
export interface FastifyTypeProvider {
  readonly schema: unknown,
  readonly input: unknown,
  readonly output: unknown,
}
```

Solution 2:
```
export interface FastifySeparatedTypeProvider {
  readonly validator: FastifyTypeProvider,
  readonly serializer: FastifyTypeProvider,
}
```

I prefer solution 1, but this requires creating breaking changes. So I created this PR https://github.com/fastify/fastify/pull/5315 to implement solution 2. What do you guys think?

Anyone could quickly create PR's to existing type providers to adapt to the new type provider interface. For users using fastify-type-provider-xxx there should be no code changes required. Should I create a PR for solution 1 aswell?

**Reply by mcollina** [reply]
If we want to do breaking changes, we are loosely working on fastify v5, which might be a good idea to do option 1.

**Reply by Bram-dc** [reply]
I have created a new PR with the breaking changes: https://github.com/fastify/fastify/pull/5427
